internal/provider/resource_user.go

```diff
@@ -5,143 +5,242 @@ import (
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
 
-	var diags diag.Diagnostics
+	client, ok := req.ProviderData.(*juju.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *juju.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
 
-	name := d.Get("name").(string)
-	displayName := d.Get("display_name").(string)
-	password := d.Get("password").(string)
+	r.client = client
+}
+
+// Create is called when the provider must create a new resource. Config
+// and planned state values should be read from the
+// CreateRequest and new state values set on the CreateResponse.
+func (r *userResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data userResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	_, err := client.Users.CreateUser(juju.CreateUserInput{
-		Name:        name,
-		DisplayName: displayName,
-		Password:    password,
+	_, err := r.client.Users.CreateUser(juju.CreateUserInput{
+		Name:        data.Name.ValueString(),
+		DisplayName: data.DisplayName.ValueString(),
+		Password:    data.Password.ValueString(),
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create user, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("created user resource %q", data.Name))
 
-	d.SetId(fmt.Sprintf("user:%s", name))
-
-	return diags
+	// Save data into Terraform state
+	data.ID = types.StringValue(newIDFromUserName(data.Name.ValueString()))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 }
 
-func resourceUserRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
-
-	var diags diag.Diagnostics
+// Read is called when the provider must read resource values in order
+// to update state. Planned state values should be read from the
+// ReadRequest and new state values set on the ReadResponse.
+// Take the juju api input from the ID, it may not exist in the plan.
+// Only set optional values if they exist.
+func (r *userResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data userResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	id := strings.Split(d.Id(), ":")
-	name := id[1]
-	response, err := client.Users.ReadUser(name)
+	// If applicable, this is a great opportunity to initialize any necessary
+	// provider client data and make a call using it.
+	response, err := r.client.Users.ReadUser(userNameFromID(data.ID.ValueString()))
```

Comment:
We need a check here that `r.client` is not nil to prevent segfaults/panics (in case `Configure` wasn't called/didn't work).

Reply:
I'm wondering why none of the examples nor real work framework code does that.